udtf.go

```diff
@@ -0,0 +1,493 @@
+	// It should be set after executing the function, but never read.
+	err := chunk.initFromDuckDataChunk(output, false)
+	if err != nil {
+		errstr := C.CString(err.Error())
+		defer C.free(unsafe.Pointer(errstr))
+		C.duckdb_function_set_error(info, errstr)
+	}
+
+	row := Row{
+		chunk:      &chunk,
+		projection: instance.projection,
+	}
+
+	maxSize := C.duckdb_vector_size()
+
+	switch fun := instance.fun.(type) {
+	case RowTableSource:
+		// At the end of the loop row.r must be the index one past the last added row
+		for row.r = 0; row.r < maxSize; row.r++ {
+			nextResults, err := fun.FillRow(row)
+			if err != nil {
+				errstr := C.CString(err.Error())
+				defer C.free(unsafe.Pointer(errstr))
+				C.duckdb_function_set_error(info, errstr)
+			}
+			if !nextResults {
+				break
+			}
+		}
+	case ThreadedRowTableSource:
+		localstateHandle := *(*cgo.Handle)(C.duckdb_function_get_local_init_data(info))
+		localState := localstateHandle.Value().(pinnedValue[any]).value
+
+		// At the end of the loop row.r must be the index one past the last added row
+		for row.r = 0; row.r < maxSize; row.r++ {
+			nextResults, err := fun.FillRow(localState, row)
+			if err != nil {
+				errstr := C.CString(err.Error())
+				defer C.free(unsafe.Pointer(errstr))
+				C.duckdb_function_set_error(info, errstr)
+			}
+			if !nextResults {
+				break
+			}
+		}
+	}
+
+	// since row.r points to one past the last value, it is also the size
+	C.duckdb_data_chunk_set_size(output, row.r)
+}
+
+//export udf_chunk_callback
+func udf_chunk_callback(info C.duckdb_function_info, output C.duckdb_data_chunk) {
+	h := *(*cgo.Handle)(C.duckdb_function_get_bind_data(info))
+	instance := h.Value().(pinnedValue[tableFunctionData]).value
+
+	var chunk DataChunk
+	// false, maybe this needs to be true but I don't think we need to read the validity mask
+	// It should be set after executing the function, but never read.
+	err := chunk.initFromDuckDataChunk(output, false)
+	if err != nil {
+		errstr := C.CString(err.Error())
+		defer C.free(unsafe.Pointer(errstr))
+		C.duckdb_function_set_error(info, errstr)
+	}
+
+	switch fun := instance.fun.(type) {
+	case ChunkTableSource:
+		err := fun.FillChunk(chunk)
+		if err != nil {
+			errstr := C.CString(err.Error())
+			defer C.free(unsafe.Pointer(errstr))
+			C.duckdb_function_set_error(info, errstr)
+		}
+	case ThreadedChunkTableSource:
+		localstateHandle := *(*cgo.Handle)(C.duckdb_function_get_local_init_data(info))
+		localState := localstateHandle.Value().(pinnedValue[any]).value
+		err := fun.FillChunk(localState, chunk)
+		if err != nil {
+			errstr := C.CString(err.Error())
+			defer C.free(unsafe.Pointer(errstr))
+			C.duckdb_function_set_error(info, errstr)
+		}
+	}
+}
+
+// RegisterTableUDF registers a TableFunctionProvider to duckdb.
+// Projectionpushdown is enabled by default, and implemented transparently.
+func RegisterTableUDF[TFT TableFunction](c *sql.Conn, name string, function TFT) error {
```

Comment:
Let's return an error if the `name` is empty or the `function` is `nil`. Are there other invalid input scenarios?

Reply:
function cannot be nil, but I did add empty name